content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+
+|||codediff|rust
+  #[panic_handler]
+  fn panic(_panic: &core::panic::PanicInfo<'_>) -> ! {
+-     loop {}
++     core::arch::wasm32::unreachable()
+  }
+|||
+
+With this in place, our program compiles again. After stripping and `wasm-opt`, the binary weighs in at 168B. Minimalism wins again!
+
+## Memory Management
+
+Of course, we have given up a lot by going non-standard. Without heap allocations, there is no `Box`, no `Vec`, no `String`, nor many of the other useful things. Luckily, we can get those back without having to provide an entire operating system. 
+
+A good chunk of what `std` offers are just re-exports from `core` and from another Rust-internal crate called `alloc`. `alloc` contains everything around memory allocations and the data structures that rely on it. By importing it, we can regain access to our trusty `Vec`.
+
+```rust
+#![no_std]
+extern crate alloc;
+use alloc::vec::Vec;
+
+#[no_mangle]
+fn nth_prime(n: usize) -> usize {
+    let mut primes: Vec<usize> = Vec::new();
+    let mut current = 2;
+    while primes.len() < n {
+        if !primes.iter().any(|prime| current % prime == 0) {
+            primes.push(current);
+        }
+        current += 1;
+    }
+    primes.into_iter().last().unwrap_or(0)
+}
+
+#[panic_handler]
+fn panic(_panic: &core::panic::PanicInfo<'_>) -> ! {
+    core::arch::wasm32::unreachable()
+}
+```
+
+Trying to compile this will fail of course. We haven’t actually told Rust what our memory management looks like:
+
+```
+
+$ cargo build --target=wasm32-unknown-unknown --release
+error: no global memory allocator found but one is required; link to std or add `#[global_allocator]` to a static item that implements the GlobalAlloc trait
+
+error: `#[alloc_error_handler]` function required, but not found
+
+note: use `#![feature(default_alloc_error_handler)]` for a default error handler
+```
+
+At the time of writing, in Rust 1.67, you need to provide an error handler that gets invoked when an allocation fails. In the next release, Rust 1.68,`default_alloc_error_handler` has been stabilized, which means every Rust non-standard Rust program will come with a default implementation of that error handler. If you want to provide your own error handler anyway, you can:
+
+```rust
+#[alloc_error_handler]
+fn alloc_error(_: core::alloc::Layout) -> ! {
+    core::arch::wasm32::unreachable()
+}
+```
+
+Without allocation errors handled, we should finally provide a way to do actual memory allocations. Just like in my [C to WebAssembly article][c to wasm], my custom allocator is going to be a minimal bump allocator. We statically allocate an arena that will function as our heap and keep track of where the “free area” begins. Because we are not using Wasm Threads, I am also going to ignore thread safety.
```

Comment:
It's not _that_ large, but implementing a whole allocator in the blog post seems just a bit out of scope, especially given the intricacies / complexity of the allocator APIs. The accompanying explanation helps at least someone who worked with unsafe Rust code before, but might very well be overwhelming for another reader.

I'd probably just explain what a bump allocator is and then link to [`bumpalo`](https://github.com/fitzgen/bumpalo) which is a `no_std`-compatible bump allocator, just like you already link to `wee_alloc` below.

Reply:
I might ignore this one :P Mostly because my C article also had a bump allocator. And I kinda want to demystify that thing a little.